src/app/shared/table-of-contents/table-of-contents.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  Component, Input, AfterViewInit, Inject, ElementRef, OnInit
+} from '@angular/core';
+import {DOCUMENT} from '@angular/platform-browser';
+import {Router, ActivatedRoute, Event, NavigationEnd} from '@angular/router';
+import {Observable} from 'rxjs/Observable';
+import {Subscription} from 'rxjs/Subscription';
+import 'rxjs/add/observable/fromEvent';
+import 'rxjs/add/operator/debounceTime';
+
+@Component({
+  selector: 'table-of-contents',
+  styleUrls: ['./table-of-contents.scss'],
+  template: `
+    <div *ngIf="links?.length" class="toc-container">
+      <h2>Contents</h2>
+      <ul>
+        <li *ngFor="let link of links; let i = index" class="li-{{link.type}}">
+          <a [href]="_rootUrl + '#' + link.id"
+            [class.active]="link.active">
+            {{link.name}}
+          </a>
+        </li>
+      </ul>
+    </div>
+  `
+})
+export class TableOfContents implements OnInit, AfterViewInit {
+
+  @Input() links: any[] = [];
+  @Input() container: string;
+
+  _rootUrl: string;
+  private _scrollContainer: any;
+  private _scrollSubscription: Subscription;
+  private _routeSubscription: Subscription;
+
+  private get scrollOffset(): number {
+    const { top } = this._element.nativeElement.getBoundingClientRect();
+    if (typeof this._scrollContainer.scrollTop !== 'undefined') {
+      return this._scrollContainer.scrollTop + top;
+    } else if (typeof this._scrollContainer.pageYOffset !== 'undefined') {
+      return this._scrollContainer.pageYOffset + top;
+    }
+  }
+
+  constructor(private _router: Router,
+              private _route: ActivatedRoute,
+              private _element: ElementRef,
+              @Inject(DOCUMENT) private _document: any) {
+
+    this._routeSubscription = this._router.events.subscribe((event: Event) => {
+      if (event instanceof NavigationEnd) {
+        const rootUrl = _router.url.split('#')[0];
+        if (rootUrl !== this._rootUrl) {
+          this.links.splice(0, this.links.length);
+          this.createLinks();
+          this._rootUrl = rootUrl;
+        }
+      }
+    });
+  }
+
+  ngOnInit(): void {
+    this._scrollContainer = this.container ?
+      this._document.querySelectorAll(this.container)[0] : window;
+
+    this._scrollSubscription = Observable
+      .fromEvent(this._scrollContainer, 'scroll')
+      .debounceTime(10)
+      .subscribe(this.onScroll.bind(this));
+  }
+
+  ngAfterViewInit(): void {
+    this._route.fragment.subscribe(fragment => {
+      setTimeout(() => {
```

Comment:
Completely agree. Needed a mechanism to delay the jump so the content injection can load and bootstrap the components. Open to better ideas, I'll add a comment in the meantime.